SCC-2079 Make function multithreaded

mwbenowitz opened the pull request:
This converts the main handler method for the function to a method that invokes a multi-processed pool of jobs which work on processing SHEP updates int parallel. This is a simple conversion where all the logic within each process represents the processing for a single record with the outcome being returned from the method invoked.

Hypothetically this will realize a large performance gain in the Poster as Neo4j claims to be able to handle parallel requests rather easily. But this has not been proven and we also have to make sure that the lambda is appropriately provisioned to take advantage of this.

This also adds some unit tests for the methods in the `app.rb` module

mwbenowitz committed:
SCC-2079 Make function multithreaded

This converts the main handler method for the function to a method that invokes a multi-processed pool of jobs which work on processing SHEP updates int parallel. This is a simple conversion where all the logic within each process represents the processing for a single record with the outcome being returned from the method invoked.

Hypothetically this will realize a large performance gain in the Poster as Neo4j claims to be able to handle parallel requests rather easily. But this has not been proven and we also have to make sure that the lambda is appropriately provisioned to take advantage of this.

This also adds some unit tests for the methods in the `app.rb` module

siberry commented on app.rb:
```diff
@@ -18,28 +19,32 @@ def init
   begin
     avro_data = record["kinesis"]["data"]
-  rescue KeyError => e
+  rescue *[KeyError, NoMethodError] => e
```
Cool. Haven't used that syntax before. I like it

mwbenowitz replied:
haha yeah, any new syntax like that is from me thinking "I would do it this way in python, how can I make that work in ruby?". If you haven't before check it out, they call it "splat" in ruby (you can do ** to unpack hashes too), it's pretty useful when passing arguments

siberry commented on spec/models/app_spec.rb:
```diff
@@ -0,0 +1,178 @@
+require 'spec_helper'
+
+describe "handler" do
```
This is the only file with double indenting. I am a fan of single indenting. And it seems to be more common. Not sure how much consistency matters here. Just pointing it out.

mwbenowitz replied:
ah good catch, this is just an artifact from my editor settings which are mostly for python.

I like this indenting because it encourages me not to nest code too deeply and think about my code structure but will definitely update this to match the standard

siberry reviewed (approved):
Multithreading seems like a good approach to try! I'm curious how dev and qa will handle this. Before the environment upgrade, I found those slower than my local environment. But I have not done much SHEP stuff since that upgrade.

mwbenowitz commented:
> Multithreading seems like a good approach to try! I'm curious how dev and qa will handle this. Before the environment upgrade, I found those slower than my local environment. But I have not done much SHEP stuff since that upgrade.

Since we've upgraded the performance seems to be good on dev at least, better than my machine (but I've throttled neo4j pretty severely on my machine, so maybe that's a bad comparison

mwbenowitz committed:
SCC-2079 Clean up to standardize

This fixes the indentation in the new unit test file and makes sure that the proper timeout is set for all deployments via travisCI
